### PR TITLE
State Transfer Redesign: Part I

### PR DESCRIPTION
**Pull request by PavelZaytsev**
## Overview

Description:
This PR includes the introduction of the redesign of a current state transfer.
This PR addresses the following issues:
- Avoids the “Two hop” problem by scheduling the Add, Restore and Heal workflows on the node that needs a state transfer. 
- Opens a way to easier implement different optimized batch transfer protocols on top of the existing framework. 
- Opens a way to handle a state transfer load dynamically and lazily.
- Opens a way to handle failures in many-to-one rebuilds dynamically and lazily.
- Opens a way to integrate with an upcoming GC mechanism. 
- Opens a way to integrate with an upcoming Commit Counter mechanism. 

Why should this be merged: 
A part of a state transfer redesign initiative. 

Related issue(s) (if applicable): #<number>
#1726

## Checklist (Definition of Done):

- [X] There are no TODOs left in the code
- [X] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [X] Change is covered by automated tests
- [X] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+ * A batch processor that transfers non-committed addresses one batch at a time
+ * via a replication protocol.
+ */
+@Slf4j
+public class ProtocolBatchProcessor implements StateTransferBatchProcessor {
+
+    /**
+     * Configurations for the retry logic.
+     */
+    private static final int MAX_RETRIES = 3;
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java]: please transform these constants to fields, to be able to adjust those parameters

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    private final AddressSpaceView addressSpaceView;
+
+    public ProtocolBatchProcessor(StreamLog streamLog, AddressSpaceView addressSpaceView) {
+        this.streamLog = streamLog;
+        this.addressSpaceView = addressSpaceView;
+    }
+
+    @Override
+    public CompletableFuture<BatchResult> transfer(Batch batch) {
+        CompletableFuture<Result<Long, BatchProcessorFailure>> protocolTransfer =
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java]: Better readability:
```
        return readRecords(batch.getAddresses(), 0)
                .thenApply(records -> records.flatMap(recs -> writeRecords(recs, streamLog)))
                .thenApply(result -> {
                    Result<BatchResultData, BatchProcessorFailure> newResult = result
                            .map(BatchResultData::new)
                            .mapError(e -> BatchProcessorFailure.builder().throwable(e).build());
                    return new BatchResult(newResult);
                })
                .exceptionally(failure -> {
                    BatchProcessorFailure error = BatchProcessorFailure.builder()
                            .throwable(failure)
                            .build();
                    return new BatchResult(Result.error(error));
                });
```

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+    /**
+     * Reads data entries by utilizing the replication protocol. If there are errors, retries.
+     *
+     * @param addresses The list of addresses.
+     * @return A result of reading records.
+     */
+    CompletableFuture<Result<List<LogData>, BatchProcessorFailure>> readRecords(List<Long> addresses,
+                                                                                int retries) {
+        return CompletableFuture.supplyAsync(() ->
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java]: please try to reorganize the code for better readability

**Comment by PavelZaytsev** [comment]
Closing because we decided to split it into 3 parts due to readibility and comprehension concerns.
All the comments will be addressed in the following PRs.

**Pull request closed without merging**